models/language_model.py

```diff
@@ -231,50 +261,76 @@ def _init_weights(self, module):
         elif isinstance(module, RMSNorm):
             module.weight.data.fill_(1.0)
 
-    def forward(self, x, attention_mask=None):
+    def forward(self, x, attention_mask=None, kv_cache=None, start_pos=0):
         if self.lm_use_tokens:
-            x = self.token_embedding(x) # Only embed the inputs when using tokens
-        
-        B , T, _ = x.size()
+            x = self.token_embedding(x)
+
+        # T_curr is the length of the current input sequence
+        B, T_curr, _ = x.size()
         
-        # Note: You could also cache these input embeddings if you want to avoid recomputing them
-        position_ids = torch.arange(T, device=x.device).unsqueeze(0).expand(B, -1) # Create position ids [0, 1, 2, ..., seq_len-1]
-        cos, sin = self.rotary_embd(position_ids) # Get rotary position embeddings
+        # Create position_ids for the current sequence based on start_pos
+        current_position_ids = torch.arange(start_pos, start_pos + T_curr, device=x.device).unsqueeze(0).expand(B, -1)
+        cos, sin = self.rotary_embd(current_position_ids) # Get rotary position embeddings for current tokens
+
+        # Initialize new KV cache if none provided
+        new_kv_cache_list = []
+        if kv_cache is None:
+            kv_cache = [None] * len(self.blocks)
+
+        for i, block in enumerate(self.blocks):
+            x, block_kv_cache = block(x, cos, sin, attention_mask, kv_cache[i])
+            new_kv_cache_list.append(block_kv_cache)
```

Comment:
At this point, we will have references to `kv_cache` (input KV cache) and `new_kv_cache_list` (output KV cache), i.e. references to the old and new tensors. I think we will have the same effect as above, i.e. duplicated memory.

I see two options:
- Rely on side-effects: the attention layer updates the cache object.
- Do something like `kv_cache[i] = block_kv_cache` on this line. We would still get a small memory peak, as we would have two cache objects for a given layer here. It would still be much better than having a copy for the whole cache :D

Reply:
I think the lowest hanging fruit is `kv_cahce[i]=block_kv_cache`.

Thanks for catching this.